tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+	local texmfoutput            = kpse.var_value("TEXMFOUTPUT")
+
+	---@param fname string
+	---@return string?
+	---@return string? -- error
+	find_in = function(fname)
+		local abs, err = pathlib.path_is_absolute(fname)
+		if abs == nil then return nil, err end
+		if abs then return fname end
+
+		if texmf_output_directory then
+			local p, err = pathutil.join(texmf_output_directory, fname)
+			if not p then return nil, err end
+			if kpse.in_name_ok_silent_extended(p) then return p end
+		end
+		if not texmf_output_directory then
+			if kpse.in_name_ok_silent_extended(fname) then return fname end
+		end
+		if texmfoutput then
+			local p, err = pathutil.join(texmfoutput, fname)
+			if not p then return nil, err end
+			if kpse.in_name_ok_silent_extended(p) then return p end
+		end
+
+		return fname
+	end
+end
+
+local find_out
+do
+	local texmf_output_directory = kpse.var_value("TEXMF_OUTPUT_DIRECTORY")
+	local texmfoutput            = kpse.var_value("TEXMFOUTPUT")
+
+	---@param fname string
+	---@return string?
+	---@return string?
+	find_out = function(fname)
+		local abs, err = pathlib.path_is_absolute(fname)
+		if abs == nil then return nil, err end
+		if abs then return fname end
+
+		local texmf_od
+		if texmf_output_directory then
+			local p, err = pathutil.join(texmf_output_directory, fname)
+			if not p then return nil, err end
+			texmf_od = p
+			if kpse.out_name_ok_silent_extended(p) then return p end
+		end
+		if not texmf_output_directory then
+			if kpse.out_name_ok_silent_extended(fname) then return fname end
+		end
+		if texmfoutput then
+			local p, err = pathutil.join(texmfoutput, fname)
+			if not p then return nil, err end
+			if kpse.out_name_ok_silent_extended(p) then return p end
+		end
+
+		return texmf_od or fname
+	end
+end
+
+-- setup something like a logging library
+local logging = {
+	file      = nil,
+	header    = "memoize-extract.lua: ",
+	indent    = "",
+	texindent = "",
+}
+do
+	local package_name = "memoize (texlua-based extraction)"
+	local ERROR   = {
+		latex     = function(a) return ("\\PackageError{%s}{%s}{%s}"):format(a.package_name or "", a.short or "", a.long or "") end,
+		plain     = function(a) return ("\\errhelp{%s}\\errmessage{%s: %s}"):format(a.long or "", a.package_name or "", a.short or "") end,
+		context   = function(a) return ("\\errhelp{%s}\\errmessage{%s: %s}"):format(a.long or "", a.package_name or "", a.short or "") end,
+		None      = function(a) return ("%s%s.\n%s"):format(a.header or "", a.short or "", a.long or "") end,
+	}
+
+	local WARNING = {
+		latex     = function(a) return ("\\PackageWarning{%s}{%s%s}"):format(a.package_name or "", a.texindent or "", a.text or "") end,
+		plain     = function(a) return ("\\message{%s: %s%s}"):format(a.package_name or "", a.texindent or "", a.text or "") end,
+		context   = function(a) return ("\\message{%s: %s%s}"):format(a.package_name or "", a.texindent or "", a.text or "") end,
```

Comment:
```suggestion
		context   = function(a) return ("\\writestatus{%s}{%s}"):format(a.package_name or "", a.text or "") end,
```

Reply:
I'm failing to find docs on the `\writestatus` macro. The only one I found was https://wiki.contextgarden.net/Command/writestatus . Is it possible this is ConTeXt specific?

Reply:
Correct, it's the ConTeXt version of `\PackageWarning`/`\PackageInfo`.

Reply:
Ok, but the script should also work for non-ConTeXt users. So I think we shouldn't/can't use these functions.

Reply:
But don't you only use these strings in the appropriate formats? Because `\PackageWarning` isn't defined in Plain or ConTeXt, only LaTeX. (I'm only suggesting to change the `context =` line, not the `plain =` or `latex =` lines)

Reply:
Oh you're completely correct, my bad :see_no_evil: